internal/service/ec2/transit_gateway_multicast_domain_test.go

```diff
@@ -0,0 +1,1007 @@
+		}
+
+		*multicastDomain = *domain
+
+		return nil
+	}
+}
+
+func testAccCheckTransitGatewayMulticastDomainDisappears(domain *ec2.TransitGatewayMulticastDomain) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+		id := aws.StringValue(domain.TransitGatewayMulticastDomainId)
+		input := &ec2.DeleteTransitGatewayMulticastDomainInput{
+			TransitGatewayMulticastDomainId: aws.String(id),
+		}
+
+		log.Printf("[DEBUG] Deleting EC2 Transit Gateway Multicast Domain (%s): %s", id, input)
+		if _, err := conn.DeleteTransitGatewayMulticastDomain(input); err != nil {
+			return err
+		}
+
+		return tfec2.WaitForTransitGatewayMulticastDomainDeletion(conn, id)
+	}
+}
+
+func testAccCheckTransitGatewayMulticastDomainDestroy(s *terraform.State) error {
+	conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_ec2_transit_gateway_multicast_domain" {
+			continue
+		}
+
+		id := rs.Primary.ID
+		domain, err := tfec2.DescribeTransitGatewayMulticastDomain(conn, id)
+		if tfawserr.ErrMessageContains(err, "InvalidTransitGatewayMulticastDomainId.NotFound", "") {
+			continue
+		}
+
+		if err != nil {
+			return err
+		}
+
+		if domain == nil {
+			continue
+		}
+
+		state := aws.StringValue(domain.State)
+		if state != ec2.TransitGatewayMulticastDomainStateDeleted {
+			return fmt.Errorf(
+				"EC2 Transit Gateway Multicast Domain (%s) still exists in a non-deleted (%s) state",
+				id, state)
+		}
+	}
+
+	return nil
+}
+
+func testAccCheckTransitGatewayMulticastDomainAssociations(domain *ec2.TransitGatewayMulticastDomain, count int, expected map[*ec2.TransitGatewayVpcAttachment][]*ec2.Subnet) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		conn := acctest.Provider.Meta().(*conns.AWSClient).EC2Conn
+		id := aws.StringValue(domain.TransitGatewayMulticastDomainId)
+
+		assocSet, err := tfec2.GetTransitGatewayMulticastDomainAssociations(conn, id)
+		if err != nil {
+			return err
+		}
+
+		assocLen := len(assocSet)
+		if assocLen != count {
+			return fmt.Errorf(
+				"expected %d EC2 Transit Gateway Multicast Domain assoctiations; got %d", count, assocLen)
+		}
+
+		expectedIDs := make(map[string][]string)
+		for attachment, subnets := range expected {
+			var subnetIDs []string
+			for _, subnet := range subnets {
+				subnetIDs = append(subnetIDs, aws.StringValue(subnet.SubnetId))
+			}
+			expectedIDs[aws.StringValue(attachment.TransitGatewayAttachmentId)] = subnetIDs
+		}
+
+		for _, assoc := range assocSet {
+			attachmentID := aws.StringValue(assoc.TransitGatewayAttachmentId)
+			actualSubnetID := aws.StringValue(assoc.Subnet.SubnetId)
+			subnetIDs := expectedIDs[attachmentID]
+			log.Printf("[DEBUG] Subnet IDS: %s", subnetIDs)
+			found := false
+			for _, subnetID := range subnetIDs {
+				if subnetID == actualSubnetID {
+					found = true
+					break
+				}
+			}
+
+			if !found {
```

Comment:
Assertions should not be done in the subroutines. Assertions should be done in the `testAcc` function. Use the subroutine to fetch the data and pass it by reference to the `testAcc`.